Item Ball refactor / Pluralize item names for giveitem and finditem

pkmnsnfrn opened the pull request:
# Description
![Brendan picking up an item ball](https://github.com/PokemonSanFran/pokeemerald/assets/77138753/6c651061-2733-42fc-8a71-a729c3416761)

- Refactors item ball scripts to unify them all under `Common_EventScript_FindItem` , using `trainerRange_berryTreeId` to define item and `movementRangeX` to define quantity
- Converts all existing item balls scripts to use `Common_EventScript_FindItem`
- Implicitly allows for multiple items to be picked up in the field
- Pluralizes the names of any and all items (in English) for when the player gets or finds multiple items

## Usage
### Existing Items
This branch will convert all the existing scripts in vanilla `pokeemerald` to use `Common_EventScript_FindItem`. No quantity is defined for these items, which defaults to 1.

If developers would like to convert item ball scripts that they have added to their repo, [@Bassoonian has written a python script capable of doing so](https://github.com/rh-hideout/pokeemerald-expansion/blob/f7bf1f4e320ce00d3434f337aaf796b2489e6ced/item_ball_refactor_helper.py). Developers will need to run some version of the following command in their repo:

`wget https://raw.githubusercontent.com/rh-hideout/pokeemerald-expansion/f7bf1f4e320ce00d3434f337aaf796b2489e6ced/item_ball_refactor_helper.py -O item_ball_refactor_helper.py && chmod +x item_ball_refactor_helper.py && python3 item_ball_refactor_helper.py && rm item_ball_refactor_helper.py`

### New Items
When a developer wants to add a new item ball to the overworld of their game, they can do so either via json or [Porymap](https://huderlem.github.io/porymap/manual/editing-map-events.html#object-events). Regardless of the method, the result will be the same - if the player has room in their bag, the item will be picked up, added to the inventory, and the flag for that item will be set.

#### Porymap (reccomended)
When adding a new object to a map, the object's fields must be set as follows:
* **Script**: `Common_EventScript_FindItem`  (the script used for all item balls)
* **Event Flag**: `FLAG_ITEM_HAMMERLOCKE_GREAT_BALL` (the flag used for the item in question, needs to be manually created by the developer and assigned)
* **Sight Radius / Berry Tree ID**: `ITEM_GREAT_BALL` (the constant for the item the player should receive)
* **Movement Radius X**: `10` (the amount of the item the player should receive. if this is 0, the player will receive 1.)

#### json
For the new object, add the following element to the `object_events` of the `map.json` for the map being edited.

```json
    {
      "graphics_id": "OBJ_EVENT_GFX_ITEM_BALL",
      "trainer_sight_or_berry_tree_id": "ITEM_GREAT_BALL", // the constant for the item the player should receive
      "script": "Common_EventScript_FindItem", // the script used for all item balls
      "flag": "FLAG_ITEM_HAMMERLOCKE_GREAT_BALL", // the flag used for the item in question, needs to be manually created by the developer and assigned
      "movement_range_x": 10, // the amount of the item the player should receive. if this is 0, the player will receive 1.

// all elements below this line can be freely changed by the developer.
      "x": 41,
      "y": 4,
      "elevation": 3,
      "movement_type": "MOVEMENT_TYPE_LOOK_AROUND",
      "movement_range_y": 0,
      "trainer_type": "TRAINER_TYPE_NONE",
    },
```

# Testing
## Clean Branch
You can recreate this branch by applying a patch or pulling the repo. From a clean version of expansion's upcoming, you can either:

~~### Patch~~
~~`wget https://github.com/rh-hideout/pokeemerald-expansion/files/13856855/patch-item_ball_refactor_testing.patch -O item.patch ; git apply item.patch ; rm item.patch`~~
Patches are only cool when its a single commmit or a string of commits. Now that I've merged with upcoming, the patches broken and I can't be arsed to fix it.

### Repo
`git remote add psf-expansion https://github.com/PokemonSanFran/pokeemerald-expansion/ ; git pull psf-expansion item_ball_refactor`

## Manual Tests
To recreate my testing environment, you can either:

### Download
`wget https://github.com/PokemonSanFran/pokeemerald-expansion/raw/025c822c3b003b568d14f68e8ee3abc87969cf42/pokeemerald.sav -O pokeemerald.sav ; wget https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/025c822c3b003b568d14f68e8ee3abc87969cf42/data/scripts/debug.inc -O data/scripts/debug.inc ; wget https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/025c822c3b003b568d14f68e8ee3abc87969cf42/data/maps/Route102/map.json -O data/maps/Route102/map.json`

### Manually Apply
* [data/scripts/debug.inc](https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/025c822c3b003b568d14f68e8ee3abc87969cf42/data/scripts/debug.inc)
* [data/maps/Route102/map.json](https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/025c822c3b003b568d14f68e8ee3abc87969cf42/data/maps/Route102/map.json)
* [pokeemerald.sav](https://github.com/PokemonSanFran/pokeemerald-expansion/raw/025c822c3b003b568d14f68e8ee3abc87969cf42/pokeemerald.sav)

## Verified Scenarios
Each item in the following lists goes to a timestamped YouTube link with each specific scenario being tested.

- [Pick up 6 Potions](https://youtu.be/pAZPPbHQPxQ?t=11)
- [Pick up 1 Great Ball](https://youtu.be/pAZPPbHQPxQt=16)
- [Pick up 12 Cheri Berry](https://youtu.be/pAZPPbHQPxQt=21)
- [Pick up 2 Storage Key](https://youtu.be/pAZPPbHQPxQt=25)
- [Pick up 11 TM35](https://youtu.be/pAZPPbHQPxQt=30)

~~- [Pick up an item ball where movementRangeX is defined as 255](https://youtu.be/N9cNcVoB59g?t=45)~~
- [Pick up an item ball where movementRangeX is defined as 15](https://youtu.be/pAZPPbHQPxQt=42)
- [Pick up an item ball where movementRangeX is defined as 0](https://youtu.be/pAZPPbHQPxQt=47)
- [Pick up an item ball where trainerRange_berryTreeId is defined as ITEMS_COUNT + 1](https://youtu.be/pAZPPbHQPxQt=51)

- [Pick up an existing item ball](https://youtu.be/pAZPPbHQPxQ?t=66)
- [Given 2 Max Potion](https://youtu.be/pAZPPbHQPxQ?t=80)
- [Given 4 Master Ball](https://youtu.be/pAZPPbHQPxQt=87)
- [Given 1 Cheri Berry](https://youtu.be/pAZPPbHQPxQt=92)
- [Given 99 TM42](https://youtu.be/pAZPPbHQPxQt=100)
- [Given 3 Basement Key](https://youtu.be/pAZPPbHQPxQt=148)

# Features this PR does NOT handle:
* https://github.com/rh-hideout/pokeemerald-expansion/issues/3944
* https://github.com/rh-hideout/pokeemerald-expansion/issues/3943
* https://github.com/rh-hideout/pokeemerald-expansion/issues/3947

# Credits

Please credit these people instead of me.

## [deokishisu](https://www.pokecommunity.com/member.php?u=22062)
* [Wrote the original implementation](https://github.com/pret/pokeemerald/wiki/Set-Up-Item-Balls-on-a-Map-Without-Needing-New-Scripts) of Set Up Item Balls on a Map Without Needing New Scripts

## [ShinyDragonHunter](https://github.com/ShinyDragonHunter)
* [Wrote the original tutorial](https://github.com/pret/pokeemerald/wiki/Set-Up-Item-Balls-on-a-Map-Without-Needing-New-Scripts) for Set Up Item Balls on a Map Without Needing New Scripts
* Provided the original code for `CopyItemNameHandlePlural`
* Pointed out an issue where ObjectEventTemplate needed to have a bitfield adjusted to allow for more than 15 items to be given

## [ghoulslash](https://www.pokecommunity.com/member.php?u=581824)
* [Wrote the original implementation](https://github.com/pret/pokeemerald/wiki/Plural-Giveitem) of `gText_ObtainedTheItems`, `EventScript_ObtainedItemMessage`, and `EventScript_ObtainedItem`.

## [Jaizu](https://jaizu.moe/)
* [Wrote the original implementation](https://github.com/pret/pokeemerald/wiki/Plural-Giveitem) of `gText_PlayerFoundItems`, `EventScript_FoundItemMessage`, and `EventScript_FoundItem`.

# Discord Contact Info

I am `pkmnsnfrn`. I have already started a [discussion thread](https://discord.com/channels/419213663107416084/1193434690321911828).

Bassoonian commented:
Wrote a Python script to convert your repo's finditems to the new format (and also cleans up the now unused finditem scripts, both in inc files and pory files): https://hastebin.com/share/epepoworow.python

This means that this should now qualify for merging.

AsparagusEduardo commented on data/scripts/item_ball_scripts.inc:
```diff
@@ -1,659 +1,4 @@
-MagmaHideout_3F_3R_EventScript_ItemEscapeRope::
-	finditem ITEM_ESCAPE_ROPE
+Common_EventScript_FindItem::
+    callnative GetItemBallIdAndAmountFromTemplate
+    finditem VAR_RESULT VAR_0x8009
```
Tabs instead of spaces

AsparagusEduardo commented on data/scripts/obtain_item.inc:
```diff
@@ -129,10 +136,23 @@ EventScript_PutBattlePyramidItemInBag::
 
 EventScript_FoundTMHM::
 	bufferitemnameplural STR_VAR_1, VAR_0x8004, VAR_0x8005
+    compare VAR_0x8005, 2
+    goto_if_lt EventScript_FoundTMHMMessage
+    buffernumberstring STR_VAR_3, VAR_0x8005
+    message gText_PlayerFoundTMHMs
+    goto EventScript_BufferTMHMsPocket
+    return
```
Tabs instead of spaces

pkmnsnfrn replied:
https://github.com/rh-hideout/pokeemerald-expansion/pull/3942/commits/9350a93cdf6f950be97a2665798693fa05946cd6

AsparagusEduardo commented on data/scripts/obtain_item.inc:
```diff
@@ -129,10 +136,23 @@ EventScript_PutBattlePyramidItemInBag::
 	return
 
 EventScript_FoundItem::
+    compare VAR_0x8001, TRUE
+    goto_if_eq EventScript_FoundItemMessage
+    buffernumberstring 0, VAR_0x8001
+    message gText_PlayerFoundItems
+    return
```
Tabs instead of spaces

pkmnsnfrn replied:
https://github.com/rh-hideout/pokeemerald-expansion/pull/3942/commits/9350a93cdf6f950be97a2665798693fa05946cd6

AsparagusEduardo commented on data/text/obtain_item.inc:
```diff
@@ -29,3 +29,9 @@ gText_NoRoomLeftForAnother::
 	.string "to the PC.$"
+
+gText_ObtainedTheItems::
+    .string "Obtained {STR_VAR_1} {STR_VAR_2}!$"
+
+gText_PlayerFoundItems::
+    .string "{PLAYER} found {STR_VAR_1} {STR_VAR_2}!$"
```
Tabs instead of spaces

pkmnsnfrn replied:
https://github.com/rh-hideout/pokeemerald-expansion/pull/3942/commits/9350a93cdf6f950be97a2665798693fa05946cd6

AsparagusEduardo commented on include/global.fieldmap.h:
```diff
@@ -73,9 +73,9 @@ struct ObjectEventTemplate
     /*0x09*/ u8 movementType;
-    /*0x0A*/ u16 movementRangeX:4;
+    /*0x0A*/ u16 movementRangeX:8; // Increased to 8 bits in expansion to allow GetItemBallIdFromTemplate to get more than 15 of an item
              u16 movementRangeY:4;
-             //u16 padding2:8;
+             //u16 padding2:4;
     /*0x0C*/ u16 trainerType;
```
This doesn't look retro-compatible.

pkmnsnfrn replied:
I'm not sure what you mean by "retro-compatible". I think ultimately, it means you are suggesting one of the following

1) there is a better way to make this change
2) we should not make this change, `Common_EventScript_FindItem` should be capped at a quantity of 15
3) we should not make this change because a different member of the struct should be used
4) something else that I'm not thinking of

mrgriffin replied:
As far as I remember, you're right about `kind`... My only concern with that one would be to check whether it ends up being used by the soon-to-be-PRed OW expansion (I think that's in one of merrp's branches?), I suspect one of these padding bytes ends up getting used, but I'm not sure which one!

pkmnsnfrn replied:
https://discord.com/channels/419213663107416084/1193434690321911828/1194358990411411556

pkmnsnfrn replied:
https://github.com/rh-hideout/pokeemerald-expansion/pull/3942/commits/bb192693a6b9282fda449facddd6d3132d3b7902

pkmnsnfrn commented:
* Pushed option 2 as discussed
* Updated Manual Tests section (RIP patches) with new json / save / debug files
* Retested Verified Scenarios and updated YouTube link
